src/server/editorServices.ts

```diff
@@ -502,7 +522,11 @@ namespace ts.server {
             return scriptInfo && scriptInfo.getDefaultProject();
         }
 
-        private ensureInferredProjectsUpToDate() {
+        /**
+         * Ensures the project structures are upto date
+         * @param refreshInferredProjects when true updates the inferred projects even if there is no pending work
+         */
+        private ensureInferredProjectsUpToDate(refreshInferredProjects?: boolean) {
             if (this.changedFiles) {
                 let projectsToUpdate: Project[];

```

Comment:
This is just `const projectsToUpdate = flatMap(this.changedFiles, f => f.containingProjects);`. If you're worried about perf we could have a version of `flatMap` that avoids allocations in the simplest case.
In the case where there are multiple changed files, `flatMap` should be more efficient than calling `concat` in a loop.
But do we need to ensure that contains no duplicates, or will `updateProjectGraphs` do that?
Also, the name of the method indicates that this is for inferred projects, but won't `changedFiles` occasionally be for non-inferred projects?